### PR TITLE
impl(GCS+gRPC): tracing connection for Rewriter

### PR DESCRIPTION
**Pull request by coryan**
Part of the changes for #9130

- - -
This change is [Reviewable](https://reviewable.io/reviews/googleapis/google-cloud-cpp/13449)

### CODE REVIEW

File: google/cloud/storage/internal/async/rewriter_connection_tracing.cc [hunk 1]
```diff
@@ -0,0 +1,92 @@
...
+namespace storage_internal {
+GOOGLE_CLOUD_CPP_INLINE_NAMESPACE_BEGIN
+
+#ifdef GOOGLE_CLOUD_CPP_HAVE_OPENTELEMETRY
+
+namespace {
+
+using google::cloud::internal::EndSpan;
+
+class AsyncRewriterConnectionTracing
```
[the remark this answers is not in the document]
Comment (coryan) [reply google/cloud/storage/internal/async/rewriter_connection_tracing.cc]: Done.

File: google/cloud/storage/internal/async/rewriter_connection_tracing.cc [hunk 2]
```diff
@@ -0,0 +1,92 @@
...
+
+}  // namespace
+
+std::shared_ptr<storage_experimental::AsyncRewriterConnection>
+MakeTracingAsyncRewriterConnection(
+    std::shared_ptr<storage_experimental::AsyncRewriterConnection> impl,
+    bool enabled) {
+  if (!enabled) return impl;
+  auto span = internal::MakeSpan("storage::AsyncConnection::RewriteObject");
+  internal::OTelScope scope(span);
```
Comment (dbolduc) [inline google/cloud/storage/internal/async/rewriter_connection_tracing.cc:73]: This scope isn't doing anything.

File: google/cloud/storage/internal/async/rewriter_connection_tracing_test.cc [hunk 3]
```diff
@@ -0,0 +1,147 @@
...
+}
+
+TEST(RewriterTracingConnection, Basic) {
+  auto span_catcher = InstallSpanCatcher();
+  AsyncSequencer<void> sequencer;
+
+  auto mock = std::make_unique<MockAsyncRewriterConnection>();
+  EXPECT_CALL(*mock, Iterate)
+      .WillOnce([&] {
+        EXPECT_TRUE(ThereIsAnActiveSpan());
```
Comment (dbolduc) [inline google/cloud/storage/internal/async/rewriter_connection_tracing_test.cc:63]: I think we would also like to `EXPECT_TRUE(OTelContextCaptured())` in all of these.

File: google/cloud/storage/internal/async/rewriter_connection_tracing_test.cc [hunk 4]
```diff
@@ -0,0 +1,147 @@
...
+        return sequencer.PushBack("Iterate(2)").then([](auto) {
+          return make_status_or(
+              RewriteObjectResponse{1000, 3000, "test-token", absl::nullopt});
+        });
+      })
+      .WillOnce([&] {
+        EXPECT_TRUE(ThereIsAnActiveSpan());
+        return sequencer.PushBack("Iterate(3)").then([](auto) {
+          return make_status_or(
```
Comment (dbolduc) [inline google/cloud/storage/internal/async/rewriter_connection_tracing_test.cc:78]: I think we want to verify the span is not active in these continuation:

```cc
        return sequencer.PushBack("Iterate(3)").then([](auto) {
          EXPECT_FALSE(ThereIsAnActiveSpan());
          EXPECT_FALSE(OTelContextCaptured());
          return ...;
        });
```

e.g. https://github.com/googleapis/google-cloud-cpp/blob/5a1fe557ae3c37e0611322099dde29e4bfd0d730/google/cloud/storage/internal/async/connection_tracing_test.cc#L96-L97

File: google/cloud/storage/internal/async/rewriter_connection_tracing.cc [hunk 5]
```diff
@@ -0,0 +1,92 @@
...
+      std::shared_ptr<storage_experimental::AsyncRewriterConnection> impl,
+      opentelemetry::nostd::shared_ptr<opentelemetry::trace::Span> span)
+      : impl_(std::move(impl)), span_(std::move(span)) {}
+  ~AsyncRewriterTracingConnection() override { EndSpan(*span_, Status{}); }
+
+  future<StatusOr<storage_experimental::RewriteObjectResponse>> Iterate()
+      override {
+    internal::OTelScope scope(span_);
+    return impl_->Iterate().then([span = span_](auto f) {
+      auto r = f.get();
```
Comment (dbolduc) [inline google/cloud/storage/internal/async/rewriter_connection_tracing.cc:43]: I don't think we want the span active in continuations of the returned future. `DetachOTelContext(...)` is how we do this. e.g.: 
https://github.com/googleapis/google-cloud-cpp/blob/5a1fe557ae3c37e0611322099dde29e4bfd0d730/google/cloud/storage/internal/async/connection_tracing.cc#L72-L72

The order of `DetachOTelContext()` and `EndSpan()` doesn't matter, so something like:
```cc
    return impl_->Iterate().then([span = span_, oc = opentelemetry::context::RuntimeContext::GetCurrent()](auto f) {
      auto r = f.get();
      internal::DetachOTelContext(oc);
      // etc...
```

**Review by dbolduc — CHANGES_REQUESTED** [review]

File: google/cloud/storage/internal/async/rewriter_connection_tracing.cc (hunk 2 above)
Comment (coryan) [reply google/cloud/storage/internal/async/rewriter_connection_tracing.cc]: Removed

File: google/cloud/storage/internal/async/rewriter_connection_tracing_test.cc (hunk 3 above)
Comment (coryan) [reply google/cloud/storage/internal/async/rewriter_connection_tracing_test.cc:64]: Done

File: google/cloud/storage/internal/async/rewriter_connection_tracing_test.cc (hunk 4 above)
Comment (coryan) [reply google/cloud/storage/internal/async/rewriter_connection_tracing_test.cc:88]: Done

File: google/cloud/storage/internal/async/rewriter_connection_tracing.cc (hunk 5 above)
Comment (coryan) [reply google/cloud/storage/internal/async/rewriter_connection_tracing.cc]: Doen

**Review by dbolduc — APPROVED** [review]

**Pull request merged**